src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+            // by using a replacement function (x) => `{ Prop: x }`.
+            private BoundPattern PullBinaryPatternsOut<TArgs>(BoundPattern pattern, TypeSymbol newInputType, ReplaceLeafPattern<TArgs> replaceLeaf, TArgs args)
+            {
+                if (pattern is not BoundBinaryPattern binaryPattern)
+                {
+                    BoundPattern result = replaceLeaf(pattern, inputType: newInputType, args);
+                    if (pattern.WasCompilerGenerated)
+                    {
+                        result = result.MakeCompilerGenerated();
+                    }
+
+                    return result;
+                }
+
+                BoundPattern resultLeft = PullBinaryPatternsOut(binaryPattern.Left, newInputType, replaceLeaf, args);
+
+                var rightInputType = binaryPattern.Disjunction ? newInputType : resultLeft.NarrowedType;
+                BoundPattern resultRight = PullBinaryPatternsOut(binaryPattern.Right, rightInputType, replaceLeaf, args);
+
+                TypeSymbol narrowedTypeForBinary = NarrowedTypeForBinary(resultLeft, resultRight, binaryPattern.Disjunction);
+
+                return new BoundBinaryPattern(binaryPattern.Syntax, binaryPattern.Disjunction,
+                    left: resultLeft,
+                    right: resultRight,
+                    newInputType, narrowedTypeForBinary);
+            }
+
+            public override BoundNode? VisitPropertySubpattern(BoundPropertySubpattern node)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            public override BoundNode? VisitPositionalSubpattern(BoundPositionalSubpattern node)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            [return: NotNullIfNotNull(nameof(pattern))]
+            BoundPattern? VisitWithInputTypeFromPattern(BoundPattern? pattern)
+            {
+                if (pattern is null)
+                {
+                    return null;
+                }
+
+                return VisitWithInputType(pattern, pattern.InputType);
+            }
+
+            BoundPattern VisitWithInputType(BoundPattern pattern, TypeSymbol inputType)
+            {
+                var savedInputType = _inputType;
+                _inputType = inputType;
+                var result = (BoundPattern)Visit(pattern);
+                _inputType = savedInputType;
+                return result;
+            }
+
+            private BoundPattern MakeBinary(bool disjunction, ArrayBuilder<BoundPattern> builder, SyntaxNode syntax)
+            {
+                if (builder.Count == 0)
+                {
+                    BoundDiscardPattern discard = MakeDiscardPattern(syntax, _inputType).MakeCompilerGenerated();
+                    if (disjunction)
+                    {
+                        // The caller normally wants to make an `... or ...` sequence but all the parts could be skipped (discards)
```

Comment:
>all the parts could be skipped (discards) [](http://example.com/codeflow?start=90&length=41)

It looks like the comment is not accurate, ```CanSkip``` skips (negated discards) for disjunction and (discards) for conjunction. Therefore, the logic here looks suspicious. In fact, if swap the branches, not a single test fails.